build/result.go

```diff
@@ -22,166 +20,246 @@ import (
-			case statusChan <- s:
-			case <-ctx.Done():
+		}()
+
+		var res *gateway.Result
+		resp, err := cc.Build(ctx, opt, product, func(ctx context.Context, c gateway.Client) (*gateway.Result, error) {
+			var err error
+			res, err = buildFunc(ctx, c)
+
+			if res != nil && err == nil {
+				// force evaluation of all targets
+				def, err2 := getDefinition(ctx, res)
+				if err2 != nil {
+					return nil, err2
+				}
+				res, err = evalDefinition(ctx, c, def)
 			}
-		}
-	}()
 
-	// get result
-	resultCtxCh := make(chan *ResultContext)
-	errCh := make(chan error)
-	go func() {
-		solveOpt := solveOpt
-		solveOpt.Ref = ""
-		buildDoneCh := make(chan struct{})
-		_, err := c.Build(context.Background(), solveOpt, "buildx", func(ctx context.Context, c gateway.Client) (*gateway.Result, error) {
-			doneErr := errors.Errorf("done")
-			ctx, cancel := context.WithCancelCause(ctx)
-			defer cancel(doneErr)
-
-			// force evaluation of all targets in parallel
-			results := make(map[*pb.Definition]*gateway.Result)
-			resultsMu := sync.Mutex{}
-			eg, egCtx := errgroup.WithContext(ctx)
-			targets.EachRef(func(def *pb.Definition) error {
-				eg.Go(func() error {
-					res2, err := c.Solve(egCtx, gateway.SolveRequest{
-						Evaluate:   true,
-						Definition: def,
-					})
-					if err != nil {
-						return err
-					}
-					resultsMu.Lock()
-					results[def] = res2
-					resultsMu.Unlock()
-					return nil
-				})
-				return nil
-			})
-			resultCtx := ResultContext{}
-			if err := eg.Wait(); err != nil {
+			if err != nil {
 				var se *errdefs.SolveError
 				if errors.As(err, &se) {
-					resultCtx.solveErr = se
-				} else {
-					return nil, err
+					resCtx := &ResultContext{
+						done:     make(chan struct{}),
+						solveErr: se,
+						gwClient: c,
+						gwCtx:    ctx,
+					}
+					close(done)
+					close(baseCh)
+					resCtxCh <- resCtx
+					errCh <- err
+					select {
+					case <-resCtx.done:
+					case <-ctx.Done():
+					}
 				}
 			}
-			res2, _ := result.ConvertResult(targets, func(def *pb.Definition) (gateway.Reference, error) {
-				if res, ok := results[def]; ok {
-					return res.Ref, nil
-				}
-				return nil, nil
-			})
+			return res, err
+		}, ch)
+		select {
+		case <-done:
+			cancel()
+			return
+		default:
+		}
+		close(baseCh)
+
+		if resp != nil {
+			respCh <- resp
+		}
+		if err != nil {
+			resCtxCh <- nil
+			errCh <- err
+			cancel()
+			return
+		}
+
+		go func() {
+			def, err := getDefinition(ctx, res)
+			if err != nil {
+				resCtxCh <- nil
+				errCh <- err
+				cancel()
+				return
+			}
 
-			// Record the client and ctx as well so that containers can be created from the SolveError.
-			resultCtx.res = res2
-			resultCtx.gwClient = c
-			resultCtx.gwCtx = ctx
-			resultCtx.gwDone = func() {
-				cancel(doneErr)
-				// wait for Build() completion(or timeout) to ensure the Build's finalizing and avoiding an error "context canceled"
+			// NOTE: ideally this second connection should be lazily opened
```

Comment:
The comments this function needs is an overall summary of what going on. A new reader has no idea why there would be a "second connection" with another `Build()` call.

Reply:
Have added *a lot* more comments, importantly, explaining the two different scenarios.

1. The build fails, from some solve error, and we preserve this gateway session.
2. The build succeeds, we close the gateway session, and open a new one (without this, the previous gateway session remains open, so we never run the exporter, which could easily have it's own collection of errors, and would require significant changes to the API)